app/Http/Controllers/API/v1/AuthController.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace App\Http\Controllers\API\v1;
+
+use App\Http\Controllers\API\ResponseController as ResponseController;
+use App\Http\Controllers\Controller;
+use App\Models\User;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Hash;
+use Illuminate\Support\Facades\Validator;
+use Psy\Util\Json;
+
+class AuthController extends ResponseController
+{
+    /**
+     * @param Request $request
+     * @return JsonResponse
+     * @api v1
+     * @todo refactor this
+     */
+    public function register(Request $request):JsonResponse {
+        $validator = Validator::make($request->all(), [
+            'username' => 'required|string|unique:users',
+            'name'     => 'required|string|',
+            'email'    => 'required|string|email|unique:users',
+            'password' => 'required|min:3|confirmed',
+        ]);
+        if ($validator->fails()) {
+            return response()->json([
+                                        'status' => 'error',
+                                        'errors' => $validator->errors()
+                                    ], 422);
+        }
+        $input             = $request->all();
+        $input['password'] = Hash::make($input['password']);
+        $user              = User::create($input);
+
+        if ($user) {
+            $userToken = $user->createToken('token');
+            return $this->sendResponse([
+                                           'token'      => $userToken->accessToken,
+                                           'message'    => 'Registration successful.',
+                                           'expires_at' => $userToken->token->expires_at->toIso8601String()
+                                       ]);
+        }
+        $error = "Sorry! Registration is not successful.";
+        return $this->sendError($error, 401);
+    }
+
+    /**
+     * Login
+     * @param Request $request
+     * @return JsonResponse
+     * @api v1
+     */
+    public function login(Request $request): JsonResponse {
+        $validator = Validator::make($request->all(), [
+            'email'    => 'required|string|email',
+            'password' => 'required'
+        ]);
+
+        if ($validator->fails()) {
+            return $this->sendError($validator->errors(), 400);
+        }
+
+        $credentials = $request->only('email', 'password');
```

Comment:
Can you please use the variables generated by the validator for security reasons? I know, in this case, no direct reason exists. But I would rather use the same procedure everywhere to be on the safe side.

`$validator->validate()`

Reply:
We use `$validator->validate()` once in the entire project, the predominant one is `Validator::make`. `$validator->validate()` also issues a HTTP redirect response, which is indesireable in an API environment. :D

The `$request->only`is, to only return email and password as an array and drop all other $request-elements. AFAIK I can't do that on a Validator-Object. At least I haven't found a working solution.